Mathlib/LinearAlgebra/JordanChevalley.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2024 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.Algebra.Squarefree.UniqueFactorizationDomain
+import Mathlib.Dynamics.Newton
+import Mathlib.FieldTheory.Perfect
+import Mathlib.LinearAlgebra.Semisimple
+
+/-!
+# Jordan-Chevalley-Dunford decomposition
+
+Given a finite-dimensional linear endomorphism `f`, the Jordan-Chevalley-Dunford theorem provides a
+sufficient condition for there to exist a nilpotent endomorphism `n` and a semisimple endomorphism
+`s`, such that `f = n + s` and both `n` and `s` are polynomial expressions in `f`.
+
+The condition is that there exists a separable polynomial `P` such that the endomorphism `P(f)` is
+nilpotent. This condition is always satisfied when the coefficients are a perfect field.
+
+The proof given here uses Newton's method and is taken from Chambert-Loir's notes:
+[Algebre](http://webusers.imj-prg.fr/~antoine.chambert-loir/enseignement/2022-23/agreg/algebre.pdf)
+
+## Main definitions / results:
+
+ * `Module.End.exists_isNilpotent_isSemisimple`: an endomorphism of a finite-dimensional vector
+   space over a perfect field may be written as a sum of nilpotent and semisimple endomorphisms.
+
+## TODO
+
+ * Uniqueness of decomposition (once we prove that the sum of commuting semisimple endomorphims is
+   semisimple, this will follow from `Module.End.eq_zero_of_isNilpotent_isSemisimple`).
+
+-/
+
+open Algebra Polynomial
+
+namespace Module.End
+
+variable {K V : Type*} [Field K] [AddCommGroup V] [Module K V] [FiniteDimensional K V] {f : End K V}
+
+theorem exists_isNilpotent_isSemisimple_of_separable_of_dvd_pow {P : K[X]} {n : ℕ}
+    (sep : P.Separable) (nil : minpoly K f ∣ P ^ n) :
+    ∃ s ∈ adjoin K {f}, IsNilpotent (f - s) ∧ s.IsSemisimple := by
```

Comment:
I think I would prefer a separate variable instead of `f - s`.
The statement would become a bit longer, but when used with `obtain` the `f = N + s` clause can often be deconstructed using `rfl`, and the resulting goal state in applications might be closer to what the user wants.

Reply:
OK done.

What you propose also has the advantage that it saves the user having to prove that nilpotent part is a also a polynomial in `f` (if they want this).

FWIW if / when we add the uniqueness my plan would be to add some API that obviates the need to use these lemmas directly along the lines of:
```lean
def semisimpleComponent [PerfectField K] : End K V := ...

def nilpotentComponent [PerfectField K] : End K V := ...
```